File: GGAndroid_2.0/GGControls/GGViewModel/src/main/java/io/ganguo/viewmodel/common/HFRecyclerViewModel.java
package io.ganguo.viewmodel.common;

import android.databinding.ViewDataBinding;
import android.view.ViewGroup;

import io.ganguo.library.ui.adapter.v7.LoadMoreListener;
import io.ganguo.viewmodel.R;
import io.ganguo.viewmodel.common.base.BaseHFRViewModel;
import io.ganguo.viewmodel.callback.ICenterLoadingView;
import io.ganguo.library.ui.view.ViewInterface;
import io.ganguo.viewmodel.databinding.IncludeHfRecyclerBinding;

/**
 * Header Footer RecyclerView 组合ViewModel (不带下拉刷新)
 * Created by Roger on 6/22/16.
 */
public abstract class HFRecyclerViewModel<T extends ViewInterface<IncludeHfRecyclerBinding>> extends BaseHFRViewModel<IncludeHfRecyclerBinding, T>
        implements LoadMoreListener {


    @Override
    public ViewGroup getHeaderContainer() {
        return getView().getBinding().llyHeader;
    }

    @Override
    public ViewGroup getFooterContainer() {
        return getView().getBinding().llyFooter;
    }

    @Override
    public ViewGroup getEmptyContainer() {
        return getView().getBinding().llyEmpty;
    }

    @Override
    public ViewGroup getLoadingContainer() {
        return getView().getBinding().flyLoading;
    }

    @Override
    public ICenterLoadingView createLoadingView() {
        return new CommonLoadingVModel();
    }

    @Override
    public ViewDataBinding getRecyclerRootView() {
        return getView().getBinding().includeRecycler;
    }

    @Override
    public int getItemLayoutId() {
        return R.layout.include_hf_recycler;
    }

}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/pack/common/HFRecyclerViewModel.kt
package io.ganguo.viewmodel.pack.common

import android.view.ViewGroup
import androidx.databinding.ViewDataBinding
import io.ganguo.viewmodel.core.viewinterface.ViewInterface
import io.ganguo.viewmodel.R
import io.ganguo.viewmodel.pack.common.base.BaseHFRViewModel
import io.ganguo.viewmodel.databinding.IncludeHfRecyclerBinding

/**
 * Header Footer RecyclerView 组合ViewModel (不带下拉刷新)
 * Created by Roger on 6/22/16.
 */
abstract class HFRecyclerViewModel<T : ViewInterface<IncludeHfRecyclerBinding>> : BaseHFRViewModel<IncludeHfRecyclerBinding, T>() {
    override val layoutId: Int by lazy {
        R.layout.include_hf_recycler
    }
    override val headerContainer: ViewGroup by lazy {
        viewInterface.binding.llyHeader
    }

    override val footerContainer: ViewGroup by lazy {
        viewInterface.binding.llyFooter
    }

    override val statusContainer: ViewGroup by lazy {
        viewInterface.binding.llyState
    }

    override val recyclerRootView: ViewDataBinding by lazy {
        viewInterface.binding.includeRecycler
    }

}